Chummer/Backend/Derived/SustainedComplexForm.cs

```diff
@@ -0,0 +1,184 @@
+/*  This file is part of Chummer5a.
+ *
+ *  Chummer5a is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Chummer5a is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Chummer5a.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  You can obtain the full source code for Chummer5a at
+ *  https://github.com/chummer5a/chummer5a
+ */
+
+using Chummer.Backend.Skills;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Drawing;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Windows.Forms;
+using System.Xml;
+using NLog;
+
+namespace Chummer
+{
+    /// <summary>
+    /// A Sustained Complex Form
+    /// </summary>
+    
+    [HubClassTag("SourceID", true, "Name", "Extra")]
+    [DebuggerDisplay("{DisplayName(GlobalOptions.DefaultLanguage)}")]
+    public class SustainedComplexForm : ComplexForm, ISustainable, INotifyPropertyChanged
+    {
+        private Guid _guiID;
+        private Guid _guiSourceID = Guid.Empty;
+        private string _strName = string.Empty;
+        private bool _blnSelfSustained = true;
+        private int _intForce = 0;
+        private int _intNetHits = 0;
+        private readonly Character _objCharacter;
+
+        public event PropertyChangedEventHandler PropertyChanged;
+
+        #region Constructor, Create, Save, Load, and Print Methods
+        public SustainedComplexForm(Character objCharacter) : base(objCharacter)
+        {
+            //Create the GUID for new sustained Complex Forms
+            _guiID = Guid.NewGuid();
+            _objCharacter = objCharacter;
+
+            PropertyChanged += OnSustainedChanged;
```

Comment:
Careful with this assignment. Because `OnSustinedChanged` refreshes character's sustaining penalties every time a property of this `SustainedComplexForm` is changed, things are going to go haywire whenever `CreateByMapping` is called,

Reply:
So putting a breakpoint on `OnSustainedChanged()` shows it isn't called, when `CreateByMapping()` was called at the moment. Because the only property that fires(?) `OnPropertyChanged()` is `SelfSustained`, which isn't mapped over.

But to future prove it, I could add an `isCreating` bool and return `OnSustainedChanged` if that is true.

Reply:
It would probably be easier to just move the `PropertyChanged` addition into `CreateByMapping` instead of having it in the constructor.

Reply:
But not all instances of the class are created by mapping.
If you load a character they are just loaded from the `.xml`

Reply:
Hmm... well, you could just *not* use events, or at least, not here. One option could be to assign refreshing of sustaining penalties on the `Character` side: every time an object is added to the appropriate `ObservableCollection`, it gets subscribed to an event where a change in `SelfSustained` causes the character to refresh their sustaining penalties. Another option could be to just trigger the given spell's character's `OnPropertyChanged` for their sustaining penalties whenever `SelfSustained` gets altered, but that's a bit hacky.

Reply:
Triggering the `Character.OnPropertyChanged()` was what I've done before.

At the moment just throwing the `Character.RefreshSustainingPenalty()` into the set accessor of `SelfSustained` would have the exact same effect, just with less code, because then I don't need any `EventHandlers` in the first place.

So it seems like there is an `ObservableCollection<T>.PropertyChanged Event` ([dotnet](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.observablecollection-1.propertychanged?view=net-5.0)) and some solutions on how to listen to property changes of items in observable collections [example](https://www.codeproject.com/Tips/694370/How-to-Listen-to-Property-Chang). I haven't got the time to really look into that right now, but from the description it sounds like an possible solution.

Another way would be to just stick the assignment at the end of `Load()` and `CreateByMapping()`, that would also do the trick. However this just calls for mistakes and annoyances, as soon as someone touches the code later on.